src/base-client.ts

```diff
@@ -1,151 +1,165 @@
+import {
+  replaceMissingListenersWithNoOps,
+  Subscription,
+  SubscriptionConstructor,
+  SubscriptionListeners,
+} from './subscription';
+import { TokenProvider } from './token-provider';
 import { createTokenProvidingStrategy } from './token-providing-subscription';
+import HttpTransport from './transport/http';
+import WebSocketTransport from './transport/websocket';
 import { createTransportStrategy } from './transports';
-import { ElementsHeaders, responseToHeadersObject } from './network';
-import { subscribeStrategyListenersFromSubscriptionListeners } from './subscribe-strategy';
-import * as PCancelable from 'p-cancelable';
 
-import WebSocketTransport from './transport/websocket';
-import HttpTransport from './transport/http';
+import PCancelable = require('p-cancelable');
 
 export interface BaseClientOptions {
-    host: string;
-    encrypted?: boolean;
-    logger?: Logger;
+  host: string;
+  encrypted?: boolean;
+  logger?: Logger;
 }
 
 export class BaseClient {
-    private host: string;
-    private XMLHttpRequest: any;
-    private logger: Logger;
-    private websocketTransport: WebSocketTransport;
-    private httpTransport: HttpTransport;
+  private host: string;
+  private XMLHttpRequest: any;
+  private logger: Logger;
+  private websocketTransport: WebSocketTransport;
+  private httpTransport: HttpTransport;
 
-    constructor(private options: BaseClientOptions) {
-        this.host = options.host.replace(/(\/)+$/, '');
-        this.logger = options.logger || new ConsoleLogger();
-
-        this.websocketTransport = new WebSocketTransport(this.host);
-        this.httpTransport = new HttpTransport(this.host);
-    }
-    
-    public request(options: RequestOptions, tokenProvider?: TokenProvider, tokenParams?: any): PCancelable {
-        if(tokenProvider){
-            return tokenProvider.fetchToken(tokenParams).then(token =>
-                {
-                    options.headers['Authorization'] = `Bearer ${token}`
-                    return executeNetworkRequest(
-                        () => this.httpTransport.request(options),
-                        options
-                    )
-                }
-            ).catch( error => {
-                console.log(error);
-            })
-        }
-        else {
-            return executeNetworkRequest(
-                () => this.httpTransport.request(options),
-                options
+  constructor(private options: BaseClientOptions) {
+    this.host = options.host.replace(/(\/)+$/, '');
+    this.logger = options.logger || new ConsoleLogger();
+    this.websocketTransport = new WebSocketTransport(this.host);
+    this.httpTransport = new HttpTransport(this.host);
+  }
+  request(
+    options: RequestOptions,
+    tokenProvider?: TokenProvider,
+    tokenParams?: any,
+  ): PCancelable.PCancelable<any> {
+    if (tokenProvider) {
+      // TODO: Unsure if this is what we want
+      return new PCancelable<any>((onCancel, resolve, reject) => {
```

Comment:
Unsure if this is how to make this work properly. 

We want to return a cancelable promise but it feels strange wrapping a promise in another promise. 

It looks like p-cancelable has some sort of utility function for wrapping existing promises to make them cancelable, which might be what we need here: https://github.com/sindresorhus/p-cancelable#pcancelablefnfn

I didn't really understand how that was working though.